cmd/ci-operator-configresolver/main.go

```diff
@@ -0,0 +1,177 @@
+	"flag"
+	"fmt"
+	"net/http"
+	"os"
+	"path/filepath"
+
+	"github.com/fsnotify/fsnotify"
+	"github.com/openshift/ci-tools/pkg/api"
+	"github.com/openshift/ci-tools/pkg/load"
+	"github.com/openshift/ci-tools/pkg/registry"
+	types "github.com/openshift/ci-tools/pkg/steps/types"
+	log "github.com/sirupsen/logrus"
+)
+
+var (
+	resolver   registry.Resolver
+	configPath string
+	regPath    string
+)
+
+func resolveConfig(w http.ResponseWriter, r *http.Request) {
+	if r.Method != "GET" {
+		w.WriteHeader(http.StatusNotImplemented)
+		w.Write([]byte(http.StatusText(http.StatusNotImplemented)))
+		return
+	}
+	query := r.URL.Query()
+	org, ok := query["org"]
+	if !ok || len(org) != 1 {
+		w.WriteHeader(http.StatusBadRequest)
+		w.Write([]byte("org query missing or incorrect"))
+		return
+	}
+	repo, ok := query["repo"]
+	if !ok || len(repo) != 1 {
+		w.WriteHeader(http.StatusBadRequest)
+		w.Write([]byte("repo query missing or incorrect"))
+		return
+	}
+	branch, ok := query["branch"]
+	if !ok || len(branch) != 1 {
+		w.WriteHeader(http.StatusBadRequest)
+		w.Write([]byte("branch query missing or incorrect"))
+		return
+	}
+	path := filepath.Join(configPath, branch[0], fmt.Sprintf("%s-%s-%s.yaml", org[0], repo[0], branch[0]))
+	config, err := load.Config(path)
+	if err != nil {
+		w.WriteHeader(http.StatusInternalServerError)
+		w.Write([]byte(fmt.Sprintf("could not load config: %v", err)))
+		return
+	}
+	if err := config.Validate(org[0], repo[0]); err != nil {
+		w.WriteHeader(http.StatusInternalServerError)
+		w.Write([]byte(fmt.Sprintf("config validation failed: %v", err)))
+		return
+	}
+	config.Tests, err = resolveConfigExternal(config.Tests)
+	if err != nil {
+		w.WriteHeader(http.StatusInternalServerError)
+		w.Write([]byte(fmt.Sprintf("config resolution failed: %v", err)))
+		return
+	}
+	jsonConfig, err := json.Marshal(config)
+	w.WriteHeader(http.StatusOK)
+	w.Write(jsonConfig)
+}
+
+func resolveConfigExternal(config []api.TestStepConfiguration) ([]api.TestStepConfiguration, error) {
+	var result []api.TestStepConfiguration
+	for _, step := range config {
+		if step.MultiStageTestConfiguration == nil {
+			result = append(result, step)
+			continue
+		}
+		newStep := api.TestStepConfiguration{
+			As:          step.As,
+			Commands:    step.Commands,
+			ArtifactDir: step.ArtifactDir,
+			Secret:      step.Secret,
+		}
+		resolvedFlow, err := resolver.Resolve(*step.MultiStageTestConfiguration)
+		if err != nil {
+			return result, err
+		}
+		multiStageConfig := api.MultiStageTestConfiguration{
+			ClusterProfile: resolvedFlow.ClusterProfile,
+			Pre:            internalStepsToAPISteps(resolvedFlow.Pre),
+			Test:           internalStepsToAPISteps(resolvedFlow.Test),
+			Post:           internalStepsToAPISteps(resolvedFlow.Post),
+		}
+		newStep.MultiStageTestConfiguration = &multiStageConfig
+		result = append(result, newStep)
+	}
+	return result, nil
+}
+
+func internalStepsToAPISteps(steps []types.TestStep) []api.TestStep {
```

Comment:
Yeah, crap. Before, we only converted one sub-structure so we only needed an internal variant of that one thing. Now, we're operating on the whole config so we need an internal structure that holds the resolved config. @bbguimaraes thoughts?

Reply:
TBH, I never understood the purpose of `pkg/steps/types.go` (and actually have been maintaining a local patch that aliases those back to the `api` types). The [only reference I've found](https://github.com/openshift/release/pull/5019) is:

> This contains some duplicate code from TestStepConfiguration as this is essentially the evolution of that struct, but I did not want to trample that code while this new config format is still in a WIP state.

so it seems we should just merge them back?

Reply:
We have two separate types and two separate things:

 1. a type that users use to configure their tests
 2. a type that ci-operator uses to run the tests

It would not be good to use a type that has possibly-nil fields, aliases, etc post-resolution, since those are syntactic sugar for the user in config.

Reply:
From a purely abstract point of view, I understand, which is why I hadn't mentioned it before.

The differences, however, are minimal: post-resolution, the `workflow` in each test and `chain` and `reference` in each step must be `nil`. Everything else is exactly the same. I think those are simple and reasonable invariants and that it's not worth it to complicate the implementation with separate types.

Reply:
I disagree. Implicit rules about which parts of a structure are or are not nil are hard to remember, challenging to learn and grow over time. k8s in general has very explicitly not chosen this path. Prow organically grew to a total chaos of fields that may or may not be set. I don't want to see us go in that direction here.